src/git_repo_iter.rs

```diff
@@ -0,0 +1,136 @@
+use std::fs;
+use std::path::{PathBuf, Path};
+use std::collections::hash_map;
+
+use crate::config::{Config, WatchConfig};
+use crate::snapshots;
+
+/// Internal structure to facilitate "recursion" without blowing up the stack. Without this, we
+/// could call self.next() recursively whenever there was an I/O error or when we reached the end
+/// of a directory listing. There's no stack space used because we just mutate GitRepoIter, so
+/// might as well turn it into a loop.
+enum CallState {
+    Yield(PathBuf),
+    Recurse,
+    Done
+}
+
+/// Iterator over all Git repos covered by a config.
+///
+/// The process is naturally recursive, traversing a directory structure, which made it a poor fit
+/// for a more typical filter/map chain.
+///
+/// Function recursion is used in a few cases:
+///  1. Errors: If we get an I/O error, we'll call self.next() again
+///  2. Empty iterator: If we get to the end of a sub-iterator, pop & start from the top
+///
+pub struct GitRepoIter<'a> {
+    config_iter: hash_map::Iter<'a, String, WatchConfig>,
+    /// A stack, because we can't use recursion with an iterator (at least not between elements)
+    sub_iter: Vec<(PathBuf, WatchConfig, fs::ReadDir)>,
+}
+
+impl<'a> GitRepoIter<'a> {
+    pub fn new(config: &'a Config) -> Self {
+        Self { config_iter: config.repos.iter(), sub_iter: Vec::new() }
+    }
+
+    fn get_next(&mut self) -> CallState {
+        // pop
+        // 
+        // Use pop here to manage the lifetime of the iterator. If we used last/peek, we would
+        // borrow a shared reference, which precludes us from borrowing as mutable when we want to
+        // use the iterator. But that means we have to return it to the vec.
+        match self.sub_iter.pop() {
+            Some((base_path, watch_config, mut dir_iter)) => {
+                let mut next_next: Option<(PathBuf, WatchConfig, fs::ReadDir)> = None;
+                let mut ret_val = CallState::Recurse;
+                let max_depth: usize = watch_config.max_depth.into();
+                if let Some(Ok(entry)) = dir_iter.next() {
+                    let child_path = entry.path();
+                    if is_valid_directory(base_path.as_path(), child_path.as_path(), &watch_config) {
+                        dbg!(child_path.as_path());
+                        if snapshots::is_repo(child_path.as_path()) {
+                            ret_val = CallState::Yield(child_path);
+                        } else if self.sub_iter.len() < max_depth {
+                            if let Ok(child_dir_iter) = fs::read_dir(child_path.as_path()) {
+                                next_next = Some((base_path.clone(), watch_config.clone(), child_dir_iter))
```

Comment:
It'd be great if these `clone`s could go, especially the `watch_config` one, since that's effectively constant & several identical copies are being added to the stack.

Reply:
I added the clone here because I need to un-move it (move it back into the `Vec`) on line 62. I suppose I could try playing around with an `Rc`, since they are indeed immutable references...

Reply:
It makes sense, I just worry about memory usage if it hits a particularly deep tree. 

An `Rc` sounds like the easier way of going about it compared to messing around with lifetimes

Reply:
Alright, changed. Try that.